jurigged/codedb.py

```diff
@@ -0,0 +1,156 @@
+import gc
+import sys
+import time
+import types
+
+from ovld import ovld
+
+##########
+# CodeDB #
+##########
+
+MAX_TIME = 0.1
+
+
+def make_audithook(self):
+    def watch_exec(event, obj):  # pragma: no cover
+        # Note: Python does not trace audit hooks, so normal use will not show
+        # coverage of this function even if it is executed
+        if event == "exec":
+            (code,) = obj
+            self.assimilate(code)
+
+    # The closure is faster than a method on _CodeDB
+    return watch_exec
+
+
+class _CodeDB:
+    def __init__(self):
+        self.codes = {}
+        self.functions = {}
+        self.last_cost = 0
+        self.always_use_cache = False
+
+    def setup(self):
+        self.collect_all()
+        sys.addaudithook(make_audithook(self))
+
+    def collect_all(self):
+        results = []
+        for obj in gc.get_objects():
+            if isinstance(obj, types.FunctionType):
+                results.append((obj, obj.__code__))
+            elif not hasattr(obj, "__getattr__"):
+                if hasattr(obj, "__conform__"):
+                    for x in gc.get_referents(obj):
+                        if isinstance(x, types.CodeType):
+                            results.append((obj, x))
+        for obj, co in results:
+            if hasattr(obj, "__qualname__"):
+                self.assimilate(
+                    co, (co.co_filename, *obj.__qualname__.split("."))
+                )
+            objects = self.functions.setdefault(co, [])
+            objects.append(obj)
+
+    def assimilate(self, code, path=()):
+        if code.co_name == "<module>":  # pragma: no cover
+            # Typically triggered by the audit hook
+            name = code.co_filename
+        elif code.co_name.startswith("<"):
+            return
+        else:
+            name = code.co_name
+        if name:
+            path = (*path, name)
+            self.codes[(*path, code.co_firstlineno)] = code
+        for ct in code.co_consts:
+            if isinstance(ct, types.CodeType):
+                self.assimilate(ct, path)
+
+    def _get_functions(self, code):
+        t = time.time()
+        results = [
+            fn
+            for fn in gc.get_referrers(code)
+            if isinstance(fn, types.FunctionType) or hasattr(fn, "__conform__")
+        ]
+        self.functions[code] = list(results)
```

Comment:
It looks like you want the return value of `_get_functions` and the cached value in `self.functions` to be separate copies so that modifications to one (e.g. `fns.remove(fn)` in `replace_code`) don't affect the other copy. But since `get_functions` may return either of those copies depending on whether there was a cache hit, and `replace_code` is always modifying the cache copy, this should lead to unpredictable and confusing behaviour.

Reply:
It did lead to unpredictable and confusing behavior when I was trying to update `ovld` yesterday, so I already have a local fix. You would have saved me half an hour here haha. Then again I didn't expect anyone would bother reviewing this so that's nice :)